cloudinit/sources/tests/test_oracle.py

```diff
@@ -418,36 +404,76 @@ class TestReadOpcMetadata:
             OPC_V1_METADATA,
         )
 
-    def test_json_decoded_value_returned(self):
+    @pytest.fixture()
+    def with_v2(self, httpretty):
+        httpretty.register_uri(
+            httpretty.GET,
+            "http://169.254.169.254/opc/v2/instance/",
+            OPC_V2_METADATA,
+        )
+
+    @pytest.fixture()
+    def no_v2(self, httpretty):
+        httpretty.register_uri(
+            httpretty.GET,
+            "http://169.254.169.254/opc/v2/instance/",
+            status=404,
+        )
+
+    def test_json_decoded_value_returned(self, with_v2):
         # read_opc_metadata should JSON decode the response and return it
+        expected = json.loads(OPC_V2_METADATA)
+        assert expected == oracle.read_opc_metadata().instance_data
+
+    @mock.patch("cloudinit.url_helper.time.sleep", lambda _: None)
+    def test_v1_fallback(self, no_v2, httpretty):
         expected = json.loads(OPC_V1_METADATA)
-        assert expected == oracle.read_opc_metadata()
+        oracle.read_opc_metadata()
+
+        assert expected == oracle.read_opc_metadata().instance_data
 
     # No need to actually wait between retries in the tests
     @mock.patch("cloudinit.url_helper.time.sleep", lambda _: None)
     @pytest.mark.parametrize(
         "failure_count,expectation",
-        [(1, does_not_raise()), (2, pytest.raises(UrlError))],
+        [
+            (1, does_not_raise()),
+            (2, does_not_raise()),
+            (3, does_not_raise()),
+            (4, pytest.raises(UrlError)),
+        ],
     )
     def test_retries(self, expectation, failure_count, httpretty):
-        responses = [httpretty.Response("", status=404)] * failure_count
-        responses.append(httpretty.Response(OPC_V1_METADATA))
+        if failure_count < 2:
+            v2_responses = [httpretty.Response("", status=404)] * failure_count
+            v2_responses.append(httpretty.Response(OPC_V2_METADATA))
+            expected = json.loads(OPC_V2_METADATA)
+        else:
+            v2_responses = [httpretty.Response("", status=404)] * 2
+            v1_responses = [httpretty.Response(
+                "", status=404)] * (failure_count - 2)
+            v1_responses.append(httpretty.Response(OPC_V1_METADATA))
+            httpretty.register_uri(
+                httpretty.GET,
+                "http://169.254.169.254/opc/v1/instance/",
+                responses=v1_responses,
+            )
+            expected = json.loads(OPC_V1_METADATA)
         httpretty.register_uri(
             httpretty.GET,
-            "http://169.254.169.254/opc/v1/instance/",
-            responses=responses,
+            "http://169.254.169.254/opc/v2/instance/",
+            responses=v2_responses,
         )
-        expected = json.loads(OPC_V1_METADATA)
         with expectation:
-            assert expected == oracle.read_opc_metadata()
+            assert expected == oracle.read_opc_metadata().instance_data
 
 
 class TestCommon_GetDataBehaviour:
```

Comment:
Since there's no actual actionable differences, this feels a bit like premature optimization to me. If there's ever a distinction we need to check for, I think we can add the updates then.